test/entity.ts

```diff
@@ -237,33 +385,269 @@ describe('entity', () => {
-        entity.decodeValueProto(valueProto),
-        expectedValue
-      );
+      describe('default `wrapNumbers: undefined`', () => {
+        it('should not wrap ints by default', () => {
+          assert.strictEqual(
+            typeof entity.decodeValueProto(valueProto),
+            'number'
+          );
+        });
+
+        it('should throw if integer value is outside of bounds', () => {
+          const expectedError = (opts: {
+            integerValue: number;
+            propertyName: string;
+          }) => {
+            return new Error(
+              'We attempted to return all of the numeric values, but ' +
+                (opts.propertyName ? opts.propertyName + ' ' : '') +
+                'value ' +
+                opts.integerValue +
+                " is out of bounds of 'Number.MAX_SAFE_INTEGER'.\n" +
+                "Please consider passing 'options.wrapNumbers=true' or\n" +
+                "'options.wrapNumbers' as\n" +
+                '{\n' +
+                '  integerTypeCastFunction: provide <your_custom_function>\n' +
+                '  properties: optionally specify property name(s) to be cutom casted' +
+                '}\n' +
+                'to prevent this error.'
+            );
+          };
+          const largeIntegerValue = Number.MAX_SAFE_INTEGER + 1;
+          const smallIntegerValue = Number.MIN_SAFE_INTEGER - 1;
+
+          const valueProto = {
+            valueType: 'integerValue',
+            integerValue: largeIntegerValue,
+            propertyName: 'phoneNumber',
+          };
+
+          const valueProto2 = {
+            valueType: 'integerValue',
+            integerValue: smallIntegerValue,
+            propertyName: 'phoneNumber',
+          };
+
+          assert.throws(() => {
+            entity.decodeValueProto(valueProto);
+          }, expectedError(valueProto));
+
+          assert.throws(() => {
+            entity.decodeValueProto(valueProto2);
+          }, expectedError(valueProto2));
+        });
+      });
+
+      describe('should wrap ints with option', () => {
+        it('should wrap ints with wrapNumbers as boolean', () => {
+          const wrapNumbers = true;
+          const stub = sinon.spy(entity, 'Int');
+
+          entity.decodeValueProto(valueProto, wrapNumbers);
+          assert.strictEqual(stub.called, true);
+        });
+
+        it('should wrap ints with wrapNumbers as object', () => {
+          const wrapNumbers = {integerTypeCastFunction: () => {}};
+          const stub = sinon.spy(entity, 'Int');
+
+          entity.decodeValueProto(valueProto, wrapNumbers);
+          assert.strictEqual(stub.called, true);
+        });
+
+        it('should call #valueOf if integerTypeCastFunction is provided', () => {
+          const stub = sinon.stub();
+          const wrapNumbers = {integerTypeCastFunction: stub};
+
+          entity.decodeValueProto(valueProto, wrapNumbers);
+          assert.strictEqual(stub.called, true);
```

Comment:
Let’s also check that the value returned is what the iTCF returned.

Reply:
added functionality to `integerTypeCastFunction`
added an assert to verify `entity.decodeValueProto` returnes the value returned by `integerTypeCastFunction`